source/core/Variants.cpp

```diff
@@ -0,0 +1,340 @@
+{
+    _ParamDef(TypeRef, VariantType);
+    _ParamDef(ErrorCluster, ErrorClust);
+    _ParamImmediateDef(StaticTypeAndData, DestData);
+    NEXT_INSTRUCTION_METHODV()
+};
+
+// Convert variant to data of given type. Error if the data types don't match
+VIREO_FUNCTION_SIGNATUREV(VariantToData, VariantToDataParamBlock)
+{
+    ErrorCluster *errPtr = _ParamPointer(ErrorClust);
+    if (!errPtr || !errPtr->status) {
+        TypeRef variantType = _Param(VariantType);
+        TypeRef destType = _ParamImmediate(DestData._paramType);
+
+        if (variantType->IsA(destType)) {
+            void* destData = _ParamImmediate(DestData)._pData;
+            variantType->CopyData(variantType->Begin(kPARead), destData);
+        } else {
+            if (errPtr) {
+                errPtr->SetError(true, 91, "Variant To Data");
+            }
+        }
+    }
+    return _NextInstruction();
+}
+
+struct SetVariantAttributesParamBlock : public VarArgInstruction
+{
+    _ParamDef(TypeRef, InputVariant);
+    _ParamDef(StringRef, Name);
+    _ParamImmediateDef(StaticTypeAndData, Value);
+    _ParamDef(Boolean, Replaced);
+    _ParamDef(ErrorCluster, ErrorClust);
+    NEXT_INSTRUCTION_METHODV()
+};
+
+VIREO_FUNCTION_SIGNATUREV(SetVariantAttribute, SetVariantAttributesParamBlock)
+{
+    ErrorCluster *errPtr = _ParamPointer(ErrorClust);
+    bool replaced = false;
+    if (!errPtr || !errPtr->status) {
+        StringRef name = _Param(Name);
+        if (IsStringEmpty(name)) {
+            if (errPtr) {
+                errPtr->SetError(true, 1, "Set Variant Attribute");
+            }
+        } else {
+            TypeRef inputVariant = _Param(InputVariant);
+            StaticTypeAndDataPointer value = &_ParamImmediate(Value);
+            VariantAttributeManager::VariantToAttributeMapType &variantToAttributeMap = VariantAttributeManager::Instance().GetVariantToAttributeMap();
+
+            auto variantToAttributeMapIter = variantToAttributeMap.find(inputVariant);
+            if (variantToAttributeMapIter != variantToAttributeMap.end()) {
+                VariantAttributeManager::AttributeMapType *attributeMap = variantToAttributeMapIter->second;
+
+                auto pairIterBool = attributeMap->insert(VariantAttributeManager::AttributeMapType::value_type(name, value));
+                replaced = !pairIterBool.second;
+                if (!pairIterBool.second) {
+                    pairIterBool.first->second = value;
+                }
+            } else {
+                auto attributeMap = new VariantAttributeManager::AttributeMapType;
+                (*attributeMap)[name] = value;
+                variantToAttributeMap[inputVariant] = attributeMap;
+                replaced = false;
+            }
+        }
+    }
+    if (_ParamPointer(Replaced)) {
+        _Param(Replaced) = replaced;
+    }
+    return _NextInstruction();
+}
+
+struct GetVariantAttributesParamBlock : public VarArgInstruction
```

Comment:
Name should be singular, returns only one attribute.

Reply:
Fixed. Thanks.